Widevine eme

realeyes-matthew opened the pull request:
Note: I can't check all the boxes on the checklist at the bottom, so any help or guidance to align this with standards is greatly appreciated

### This PR will...

...revise the EME controller to allow the user to control the EME configuration process. It exposes 3 hooks that the user implements if `emeEnabled` is true. These hooks allow the user to select the desired KeySystem (requestMediaKeySystemAccessFunc), provide the key session with the init data when a request is generated (getEMEInitializationDataFunc) and control the license retrieval flow (getEMELicenseFunc). Events and errors have been added for each step in the EME configuration process, as well as guards in the stream controllers to ensure that DRM-encrypted data is not decrypted by the key loader. API docs have also been updated.

### Why is this Pull Request needed?

Currently, the user does not have much control over the EME configuration process, especially around license requests (e.g. appending tokens). The user can now also choose the KeySystem they want to use in the requestMediaKeySystemAccessFunc, which helps us prepare to support Playready. This PR also splits out EME-related events from key events, and guards against DRM-encrypted data being decrypted like traditional AES-encrypted content.

### Are there any points in the code the reviewer needs to double check?

I need to write tests, so any help or guidance with doing this correctly would be greatly appreciated.

### Resolves issues:

This PR includes documentation per Issue #2016

### Checklist

- [X] changes have been done against master branch, and PR does not conflict
- [x] new unit / functional tests have been added (whenever applicable)
- [x] API or design changes are documented in API.md

johnBartos commented:
@realeyes-matthew Looks like you need to rebase against master

michaelcunningham19 commented on src/hls.js:
```diff
@@ -197,17 +198,6 @@ export default class Hls extends Observer {
       networkControllers.push(subtitleTrackController);
     }
 
-    Controller = config.emeController;
```
Wouldn't this change make it so the EMEController always exists? Pretty certain we omit this from the light build

realeyes-matthew replied:
Ah got it, reverting these changes

realeyes-matthew commented:
@michaelcunningham19 Thanks for the reply. I've removed the code that is not related to EME, squashed my commits and removed the changes in dist. The PR should actually now be ready for review :)

realeyes-matthew commented:
This PR is ready for review. I updated the changes to listen for the `encrypted` event, so users can now configure EME normally, and even if the PSSH box has been stripped from the init frag. API docs have been revised and the demo has been updated to reflect the new API.

realeyes-matthew commented:
@johnBartos @itsjamie I finished writing the tests for the new EME controller. However, when I tried changing the base to 1.0.0, we have all sorts of changes unrelated to EME. Do you have any advice on how I should proceed since we want this to go into 1.0.0?

johnBartos commented:
@realeyes-matthew Try again, I rebased 1.0.0 against master

realeyes-matthew commented:
@johnBartos Worked like a charm, now just needs review

itsjamie commented on src/controller/audio-stream-controller.js:
```diff
@@ -273,7 +273,7 @@ class AudioStreamController extends BaseStreamController {
         if (frag) {
           // logger.log('      loading frag ' + i +',pos/bufEnd:' + pos.toFixed(3) + '/' + bufferEnd.toFixed(3));
-          if (frag.encrypted) {
+          if (frag.encrypted && !this.hls.config.emeEnabled) {
```
Is there a better way for us to signal the style of encryption at a fragment level?

Perhaps we change frag.encrypted to be a string with the type of encryption and we set it when we create the fragment data object?

realeyes-matthew replied:
That's a possibility, I implemented it this way because detecting whether or not a stream is encrypted with DRM or simple AES encryption requires a user to configure the library accordingly. Though, having the style of encryption on the `frag` object would more robustly signal the need to (in this case) not do AES decryption if EME is enabled, even if an if check is still needed here. @itsjamie Please advise if you'd like me to do the style of encryption signaling in this PR or if I should do it as an improvement in another PR

itsjamie commented on tests/test-streams.js:
```diff
@@ -146,10 +146,45 @@ module.exports = {
+
+      return licensePromise;
+    },
+    NOTE: 'Configuring EME requires the user to implement EME hooks. See API docs for more information'
```
Any plans to show a NOTE somewhere? If not, might as well make it a comment if it's just meant to be a source comment.

realeyes-matthew replied:
I did this because I couldn't render functions in the demo config viewer, so I wanted to ensure the user knew that they had to configure EME using the hooks defined in the API docs

itsjamie commented on tests/unit/controller/eme-controller.js:
```diff
@@ -1,33 +1,103 @@
+
+const emeConfig = {
+  emeEnabled: true,
+  requestMediaKeySystemAccessFunc: function (supportedConfigurations) {
```
I think we can keep the public API breakage limited if we make these the default implementation of these hooks that can be overridden to support manifest-based loading.

Ideally, the common scenario of DRM for PlayReady and Widevine with fragment signalled encryption shouldn't require configuration from the user beyond setting the license server URLs.

itsjamie commented on tests/unit/controller/eme-controller.js:
```diff
@@ -36,71 +106,250 @@ describe('EMEController', function () {
+    });
   });
+
+  it('should request licenses for each level and audio track and apply them to their key sessions', function (done) {
```
This ideally would be true only if the license is actually different for each level, which isn't common with the streams I've worked with for DRM. Do we have a method of determining if we already have the correct license for the stream?

itsjamie commented on karma.conf.js:
```diff
@@ -72,14 +73,33 @@ module.exports = function (config) {
+    // HTTP Server options
+    // allows tester to pass certs and keys for testing EME over https
+    httpsServerOptions: process.env.KARMA_SSL ? {
+      key: fs.readFileSync('server.key', 'utf8'), // user must set the path to their SSL key
```
Is the expectation that some process out-of-band will have generated self-signed certificates before executing the unit tests with KARMA_SSL enabled?

itsjamie commented on src/controller/eme-controller.ts:
```diff
@@ -86,423 +25,373 @@ interface MediaKeysListItem {
-        });
-        return;
-      }
+    levels.forEach((level) => {
```
Is there a reason to filter levels so we don't end up with duplicate entries with the same video codec or audio codec signalled in capabilities?

itsjamie replied:
What I mean is: does EME care if we signal the same thing multiple times.

itsjamie commented on src/controller/eme-controller.ts:
```diff
@@ -86,423 +25,373 @@ interface MediaKeysListItem {
+    }).then((mediaKeys) => {
+      logger.log('Set media keys on media');
+
+      const levelRequests = this.manifestData.levels.map((level) => {
```
I'm concerned, for some playlists this seems like we could be sending upwards of 15+ license requests if you assume a full bitrate ladder with two or three audio streams. For content where the same license applies to all streams, this seems like an excessive burden to place on the license server, and on how many downstream systems it has to check with to issue the license. There must be a way for us to determine if this is needed.

itsjamie commented on src/controller/eme-controller.ts:
```diff
@@ -86,423 +25,373 @@ interface MediaKeysListItem {
+      this._media = media; // keep reference of media
+
+      this.media.addEventListener('encrypted', (event) => {
+        this.initDataType = event.initDataType;
```
Should early exit this function if user configures init data in fragment to false.

itsjamie commented on src/controller/eme-controller.ts:
```diff
@@ -86,423 +25,373 @@ interface MediaKeysListItem {
-    case KeySystems.WIDEVINE:
-      // For Widevine CDMs, the challenge is the keyMessage.
-      return keyMessage;
+  // Getters for EME Controller
```
What is hitting the controller expecting it to have a public API? I'd be wary of adding API here, since I'm pretty sure we don't normally expect users to directly access the controllers, and they're generally viewed as standalone components.

robwalch commented:
Currently all work for v1.0.0 should be done against https://github.com/video-dev/hls.js/tree/feature/v1.0.0

It would be great to see this rebased against that branch. I'm not familiar with the work in video-dev:1.0.0.

springuper commented on src/controller/eme-controller.ts:
```diff
@@ -86,423 +25,372 @@ interface MediaKeysListItem {
+  onMediaDetaching() {
+    if (this.emeEnabled) {
+      const keySessionClosePromises: Promise<void>[] = this._keySessions.map((keySession) => {
+        return keySession.close();
```
nice!

should we execute `this._media.setMediaKeys(null)` as well?
